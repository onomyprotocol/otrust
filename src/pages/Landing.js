import React from 'react';
import styled from 'styled-components';

import { AccentButton } from 'components/UI/Button';
import logo from '../assets/images/onomy.png';
import rightCursor from '../assets/images/rightCursor.png';
import metamask from '../assets/images/metamask.png';
import ledger from '../assets/images/ledger.png';
import coinbase from '../assets/images/coinbase.png';
import walletConnect from '../assets/images/walletConnect.png';
import { SUPPORTED_WALLETS } from '../connectors';

const wallets = [
  { title: 'Metamask', img: metamask },
  { title: 'Ledger', img: ledger },
  { title: 'Wallet Connect', img: walletConnect },
  { title: 'Coinbase Wallet', img: coinbase },
];

const Wrapper = styled.div`
  text-align: center;
`;

const StyledHeader = styled.header`
  background-color: ${props => props.theme.colors.bgNormal};
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  font-size: calc(10px + 2vmin);
  color: white;
`;

const StyledTopPart = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 30vh;
  background-color: ${props => props.theme.colors.bgDarken};
  width: 100%;
`;

const StyledLogo = styled.img`
  width: 120px;
  height: 95.21px;
  margin-bottom: 26px;
`;

const StyledLogoText = styled.span`
  text-align: left;
  font: normal normal bold 40px/21px 'Bebas Neue';
  letter-spacing: 1.6px;
  color: #e1dfeb;
  opacity: 1;
`;

const StyledBottomPart = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  height: 70vh;
`;

const BottomTitleText = styled.span`
  text-align: center;
  font: normal normal medium 22px/30px Poppins;
  letter-spacing: 0px;
  color: ${props => props.theme.colors.txtPrimary};
`;

const BottomDescriptionText = styled.span`
  font: normal normal normal 14px/20px Poppins;
  letter-spacing: 0px;
  color: ${props => props.theme.colors.txtSecondary};
  width: 300px;
  margin-top: 16px;
  margin-bottom: 40px;
`;

const WalletWrapper = styled(AccentButton)`
  width: 385px;
  height: 75px;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  background-color: ${props => props.theme.colors.bgDarken};
  padding: 24px;
  border-radius: 8px;
  margin-bottom: 16px;
`;

const IconWrapper = styled.div`
  width: 43px;
`;

const WalletIcon = styled.img`
  height: 27px;
`;

const WalletText = styled.span`
  font: normal normal medium 16px/25px Poppins;
  color: ${props => props.theme.colors.txtSecondary};
`;

const RightIcon = styled.img`
  width: 24px;
`;

export default function Landing({ connectWallet }) {
  const onWalletClick = wallet => {
    Object.values(SUPPORTED_WALLETS).forEach(sWallet => {
      if (sWallet.name === wallet.title) {
        if (sWallet.name === 'Injected') {
          if (typeof web3 !== 'undefined') {
            connectWallet(sWallet.connector);
          } else {
            window.open('https://metamask.io/download.html');
          }
        } else {
          connectWallet(sWallet.connector);
        }
      }
    });
  };

  return (
    <Wrapper>
      <StyledHeader>
        <StyledTopPart>
          <StyledLogo src={logo} alt="logo" />
          <StyledLogoText>ONOMY</StyledLogoText>
        </StyledTopPart>
        <StyledBottomPart>
          <BottomTitleText>Connect Your Wallet</BottomTitleText>
          <BottomDescriptionText>
            To participate bonding curve process and buy NOM tokens you need to connect your Eth wallet
          </BottomDescriptionText>
          {wallets.map(wallet => (
            <WalletWrapper key={wallet.title} onClick={() => onWalletClick(wallet)}>
              <IconWrapper>
                <WalletIcon alt={`${wallet.title} Icon`} src={wallet.img} />
              </IconWrapper>
              <WalletText>{wallet.title}</WalletText>
              <RightIcon alt="Right Cursor" src={rightCursor} />
            </WalletWrapper>
          ))}
        </StyledBottomPart>
      </StyledHeader>
    </Wrapper>
  );
}
