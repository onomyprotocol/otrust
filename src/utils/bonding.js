const a = 100000000;

export function priceAtSupply(supply) {
  return (supply / a) ** 2;
}

// Bonding curve price at specified supply
export function supplyAtPrice(price) {
  //#NOM Sold = sqrt(ETH/NOM) * a
  return Math.sqrt(price) * a;
}

// NOM supply range to ETH
export function NOMsupplyETH(supplyTop, supplyBot) {
  // Integrate over curve to get amount of ETH needed to buy amount of NOM
  // ETH = a/3((supplyNOM_Top/a)^3 - (supplyNOM_Bot/a)^3)
  return (a / 3) * ((supplyTop / a) ** 3 - (supplyBot / a) ** 3);
}

// Returns Buy Quote for the purchase of NOM based on amount of ETH
// Parameters:
// Input
// amount: amount of ETH
// Output
// output: amount of NOM
export function ETHtoNOM(amount, supply) {
  const price = priceAtSupply(supply);
  // 1. Determine supply bottom
  // Price bottom is 1% above priceBondCurve
  const priceBot = price * 1.01;
  const supplyBot = supplyAtPrice(priceBot);
  // 2. Integrate over curve, and solve for supply top
  // (3*ETH/a + (supplyNOM_Bot/a)^3)^(1/3)
  const supplyTop = a * ((3 * amount) / a + supplyBot ** 3 / a ** 3) ** (1 / 3);
  // 3. Subtract supply bottom from top to get amount NOM for amount ETH
  const diff = supplyTop - supplyBot;
  return { supplyBot, supplyTop, diff };
}

// Returns Sell Quote: NOM for ETH (Dec 18)
// Parameters:
// Input
// uint256 amountNOM: amount of NOM to be sold (18 decimal)
// Output
// uint256: amount of ETH paid in Wei or ETH (18 decimal)
export function NOMtoETH(amount, supplyTop) {
  const supplyBot = supplyTop - amount;
  const diff = 0.99 * NOMsupplyETH(supplyTop, supplyBot);
  return { supplyBot, supplyTop, diff };
}
