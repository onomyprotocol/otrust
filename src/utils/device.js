const size = {
  mobileS: '320px',
  mobileM: '375px',
  mobileL: '425px',
  tablet: '768px',
  laptop: '1024px',
  desktopS: '1366px',
  desktopM: '1920px',
  desktopL: '2560px',
};

export const device = {
  mobileS: `(min-width: ${size.mobileS})`,
  mobileM: `(min-width: ${size.mobileM})`,
  mobileL: `(min-width: ${size.mobileL})`,
  tablet: `(min-width: ${size.tablet})`,
  laptop: `(min-width: ${size.laptop})`,
  desktopS: `(min-width: ${size.desktopS})`,
  desktopM: `(min-width: ${size.desktopM})`,
  desktopL: `(min-width: ${size.desktopL})`,
};
