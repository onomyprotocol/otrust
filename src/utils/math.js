import { BigNumber } from 'bignumber.js';

export function chopFloat(value, precision) {
  return Math.round(value * 10 ** precision) / 10 ** precision;
}

export function randomInt(min, max) {
  return Math.floor(Math.random() * (max - min + 1) + min);
}

export function truncate(str, maxDecimalDigits) {
  if (str.includes('.')) {
    const parts = str.split('.');
    return parts[0] + '.' + parts[1].slice(0, maxDecimalDigits);
  }
  return str;
}

export function format18(bignumber) {
  return bignumber.div(new BigNumber(10 ** 18));
}

export function parse18(bignumber) {
  return bignumber.times(new BigNumber(10 ** 18));
}

export var isNumber = function isNumber(value) {
  return typeof value === 'number' && isFinite(value);
};
