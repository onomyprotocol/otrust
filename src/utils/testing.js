import { ThemeProvider } from 'styled-components';
import ApolloClient, { InMemoryCache } from 'apollo-boost';
import { ApolloProvider } from '@apollo/client';
import { render } from '@testing-library/react';
import { BigNumber } from 'bignumber.js';

import { darkNew } from 'Theme/theme';
import { ChainContext } from '../context/chain/ChainContext';
import { ExchangeContext } from '../context/exchange/ExchangeContext';

const client = new ApolloClient({
  uri: process.env.REACT_APP_GRAPHQL_ENDPOINT,
  cache: new InMemoryCache(),
});

export const renderWithTheme = (Component, props, children) => {
  if (children) {
    return render(
      <ThemeProvider theme={darkNew}>
        <Component {...props}>{children}</Component>
      </ThemeProvider>,
    );
  }
  return render(
    <ThemeProvider theme={darkNew}>
      <Component {...props} />
    </ThemeProvider>,
  );
};

export const renderWithContext = (Component, props) => {
  return render(
    <ThemeProvider theme={darkNew}>
      <ApolloProvider client={client}>
        <ChainContext.Provider value={{ supplyNOM: new BigNumber(0), theme: darkNew }}>
          <ExchangeContext.Provider value={{ askAmount: 0, bidAmount: 0, bidDenom: 'strong' }}>
            <Component {...props} />
          </ExchangeContext.Provider>
        </ChainContext.Provider>
      </ApolloProvider>
    </ThemeProvider>,
  );
};
