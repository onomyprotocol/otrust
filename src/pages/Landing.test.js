import { cleanup, fireEvent } from '@testing-library/react';

import Landing, { wallets } from './Landing';
import { renderWithTheme } from '../utils/testing';

const testProps = {
  connectWallet: jest.fn(),
};

describe('Given the Landing component', () => {
  describe('when the component is rendered', () => {
    afterEach(cleanup);

    it('should match the snapshot', () => {
      const { asFragment } = renderWithTheme(Landing, testProps);
      expect(asFragment()).toMatchSnapshot();
    });

    it('should include required wallets', () => {
      const { getByText } = renderWithTheme(Landing, testProps);
      expect(getByText('Metamask')).toBeInTheDocument();
      expect(getByText('Ledger')).toBeInTheDocument();
      expect(getByText('Wallet Connect')).toBeInTheDocument();
      expect(getByText('Coinbase Wallet')).toBeInTheDocument();
    });

    describe('and user clicks on Wallet title', () => {
      it('should call connectWallet function with correct params', async () => {
        const { getByText } = renderWithTheme(Landing, testProps);

        fireEvent.click(getByText(wallets[0].title));
        fireEvent.click(getByText(wallets[1].title));
        fireEvent.click(getByText(wallets[2].title));
        fireEvent.click(getByText(wallets[3].title));

        expect(testProps.connectWallet).toHaveBeenCalledTimes(4);
      });
    });
  });
});
