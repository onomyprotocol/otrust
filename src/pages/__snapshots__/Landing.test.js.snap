// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given the Landing component when the component is rendered should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-eCApnc fLcWfp"
  >
    <header
      class="sc-jSFjdj hxpqMV"
    >
      <div
        class="sc-gKAaRy ZgfvY"
      >
        <img
          alt="logo"
          class="sc-iCoGMd cLAZvB"
          src="onomy.png"
        />
        <span
          class="sc-fujyAs emMVMf"
        >
          ONOMY
        </span>
      </div>
      <div
        class="sc-pNWdM jFbpJL"
      >
        <span
          class="sc-jrsJWt gemQzB"
        >
          Connect Your Wallet
        </span>
        <span
          class="sc-kEqXSa kPIfJj"
        >
          To participate bonding curve process and buy NOM tokens you need to connect your Eth wallet
        </span>
        <button
          class="sc-gtsrHT sc-iqAclL cyZncv VYyzG"
        >
          <div
            class="sc-crzoAE bQuOpk"
          >
            <img
              alt="Metamask Icon"
              class="sc-dIsUp hgyPCw"
              src="metamask.png"
            />
          </div>
          <span
            class="sc-bqGGPW kJEmxb"
          >
            Metamask
          </span>
          <img
            alt="Right Cursor"
            class="sc-ksluID ieDcGB"
            src="rightCursor.png"
          />
        </button>
        <button
          class="sc-gtsrHT sc-iqAclL cyZncv VYyzG"
        >
          <div
            class="sc-crzoAE bQuOpk"
          >
            <img
              alt="Ledger Icon"
              class="sc-dIsUp hgyPCw"
              src="ledger.png"
            />
          </div>
          <span
            class="sc-bqGGPW kJEmxb"
          >
            Ledger
          </span>
          <img
            alt="Right Cursor"
            class="sc-ksluID ieDcGB"
            src="rightCursor.png"
          />
        </button>
        <button
          class="sc-gtsrHT sc-iqAclL cyZncv VYyzG"
        >
          <div
            class="sc-crzoAE bQuOpk"
          >
            <img
              alt="Wallet Connect Icon"
              class="sc-dIsUp hgyPCw"
              src="walletConnect.png"
            />
          </div>
          <span
            class="sc-bqGGPW kJEmxb"
          >
            Wallet Connect
          </span>
          <img
            alt="Right Cursor"
            class="sc-ksluID ieDcGB"
            src="rightCursor.png"
          />
        </button>
        <button
          class="sc-gtsrHT sc-iqAclL cyZncv VYyzG"
        >
          <div
            class="sc-crzoAE bQuOpk"
          >
            <img
              alt="Coinbase Wallet Icon"
              class="sc-dIsUp hgyPCw"
              src="coinbase.png"
            />
          </div>
          <span
            class="sc-bqGGPW kJEmxb"
          >
            Coinbase Wallet
          </span>
          <img
            alt="Right Cursor"
            class="sc-ksluID ieDcGB"
            src="rightCursor.png"
          />
        </button>
      </div>
    </header>
  </div>
</DocumentFragment>
`;
